### ISSUE TITLE
Overlap in negative sample partitioning during dataset processing

### ISSUE THREAD

**Issue by rshibin** [Zhangcx19/GPFedRec#6]
First of all, thanks for sharing your priceless code.

After carefully reading the code, I think there is a problem in the negative sample sampling process of the function for building the training set in the data.py file. The negative samples of the training set are directly sampled from ‘negative_items’, but the 198 negative samples of the test set and validation set are also sampled from ’negative_items’. This might cause the training set to see the data of the validation set in advance, thereby enhancing the false performance.So I made the following modifications to the code：

In the _sample_negative function:

        if self.isolate:
            interact_status['negative_items_train'] = interact_status.apply(lambda row: row['negative_items'] - set(row['negative_samples']), axis=1)
            return interact_status[['userId', 'negative_items', 'negative_samples', 'negative_items_train']]
        else:
            return interact_status[['userId', 'negative_items', 'negative_samples']]

'negative_items_train' is the collection of samples from the exclusion validation set that I newly added.

In the store_all_train_data function:

        neg_col = 'negative_items_train' if self.isolate else 'negative_items'
        train_ratings = pd.merge(self.train_ratings, self.negatives[['userId', neg_col]], on='userId')
        train_ratings['negatives'] = train_ratings[neg_col].apply(lambda x: random.sample(x, num_negatives))

Through the above code modifications, I conducted a comparative experiment and found that after isolating the negative sample set for sampling, the hit rate of the model dropped significantly.Although I highly appreciate the innovation of using item-embadding to construct the user graph in the paper, there seems to be something wrong with the processing of the dataset.

Thanks again for the great work!
